stock_picking_batch_print_invoices/tests/test_batch_print_invoices.py

```diff
@@ -0,0 +1,197 @@
+    @classmethod
+    def setUpClass(cls, chart_template_ref=None):
+        super().setUpClass(chart_template_ref=chart_template_ref)
+        cls.productA = cls.env["product.product"].create(
+            {
+                "name": "Product A",
+                "type": "product",
+                "invoice_policy": "order",
+            }
+        )
+        cls.productB = cls.env["product.product"].create(
+            {
+                "name": "Product B",
+                "type": "product",
+                "invoice_policy": "delivery",
+            }
+        )
+        cls.client_1 = cls.env["res.partner"].create({"name": "Client 1"})
+        cls.sale_1 = cls.env["sale.order"].create(
+            {
+                "partner_id": cls.client_1.id,
+                "order_line": [
+                    (
+                        0,
+                        0,
+                        {
+                            "product_id": cls.productA.id,
+                            "product_uom_qty": 1,
+                        },
+                    )
+                ],
+            }
+        )
+        cls.sale_1.action_confirm()
+        cls.picking_1 = cls.sale_1.picking_ids[0]
+        context = {
+            "active_model": "sale.order",
+            "active_ids": [cls.sale_1.id],
+            "active_id": cls.sale_1.id,
+            "default_journal_id": cls.company_data["default_journal_sale"].id,
+        }
+        wizard = (
+            cls.env["sale.advance.payment.inv"]
+            .with_context(**context)
+            .create(
+                {
+                    "advance_payment_method": "delivered",
+                }
+            )
+        )
+        wizard.create_invoices()
+        cls.invoice_1 = cls.sale_1.invoice_ids[0]
+        cls.invoice_1.action_post()
+        cls.client_2 = cls.env["res.partner"].create({"name": "Client 2"})
+        cls.sale_2 = cls.env["sale.order"].create(
+            {
+                "partner_id": cls.client_2.id,
+                "order_line": [
+                    (
+                        0,
+                        0,
+                        {
+                            "product_id": cls.productB.id,
+                            "product_uom_qty": 1,
+                        },
+                    )
+                ],
+            }
+        )
+        cls.sale_2.action_confirm()
+        cls.picking_2 = cls.sale_2.picking_ids[0]
+        cls.batch = cls.env["stock.picking.batch"].create(
+            {
+                "name": "Batch 1",
+                "company_id": cls.env.company.id,
+            }
+        )
+        cls.type_operation = cls.picking_1.picking_type_id
+
+    def test_stock_picking_batch_print_invoices_01(self):
+        """Batch without pickings, with any configuration in operation type,
+        batch mustn't print."""
+        for i in range(2):
+            for j in range(2):
+                for k in range(2):
+                    for m in range(2):
```

Comment:
**suggestion:** use higher-order functions to reduce cyclomatic complexity:

```suggestion
        for i, j, k, m in itertools.product(*map(range, itertools.repeat(2, 4)))
```

Please, if possible, give variables more descriptive names.